src/components/Inputs/MultiSelect.tsx

```diff
@@ -0,0 +1,169 @@
+}
+
+interface MultiSelectOptionsModalProps<T> extends MultiSelectOptionsProps<T> {
+  onClose: () => void
+}
+
+export interface MultiSelectOption<T> {
+  value: T
+  label: string
+}
+
+function MultiSelect<T>({ selectedOptions, label, renderSelectedValue, className, ...props }: MultiSelectProps<T>) {
+  const [isOptionsModalOpen, setIsOptionsModalOpen] = useState(false)
+
+  const openOptionsModal = () => setIsOptionsModalOpen(true)
+
+  return (
+    <InputArea
+      className={className}
+      onMouseDown={openOptionsModal}
+      onKeyDown={(e) => onEnterOrSpace(e, openOptionsModal)}
+    >
+      <InputLabel isElevated={selectedOptions.length > 0}>{label}</InputLabel>
+      <SelectedValue>
+        <Truncate>{renderSelectedValue()}</Truncate>
+      </SelectedValue>
+      <ModalPortal>
+        {isOptionsModalOpen && (
+          <MultiSelectOptionsModal
+            selectedOptions={selectedOptions}
+            onClose={() => setIsOptionsModalOpen(false)}
+            {...props}
+          />
+        )}
+      </ModalPortal>
+    </InputArea>
+  )
+}
+
+export function MultiSelectOptionsModal<T>({
+  options,
+  selectedOptions,
+  renderOption,
+  getOptionKey,
+  getOptionText,
+  modalTitle,
+  onClose,
+  selectedOptionsSetter
+}: MultiSelectOptionsModalProps<T>) {
+  const { t } = useTranslation()
+
+  const [focusedOptionIndex, setFocusedOptionIndex] = useState<number>()
+
+  const allOptionsAreSelected = selectedOptions.length === options.length
+
+  const handleOptionClick = (optionClicked: T) => {
+    const index = selectedOptions.findIndex((option) => option === optionClicked)
+
+    index !== -1
+      ? selectedOptionsSetter(removeItemFromArray(selectedOptions, index))
+      : selectedOptionsSetter([optionClicked, ...selectedOptions])
+  }
+
+  const handleAllButtonsClick = () => {
+    selectedOptions.length === options.length ? selectedOptionsSetter([]) : selectedOptionsSetter(options)
+  }
+
+  return (
+    <Popup
+      title={modalTitle}
+      onClose={onClose}
+      extraHeaderContent={
+        <AllButton role="secondary" short onClick={handleAllButtonsClick}>
+          {allOptionsAreSelected ? t('Unselect all') : t('Select all')}
+        </AllButton>
+      }
+    >
+      <Options>
+        {options.map((option, index) => (
+          <OptionItem
+            key={getOptionKey(option)}
+            tabIndex={0}
+            role="listitem"
+            onClick={() => handleOptionClick(option)}
+            selected={selectedOptions.some((o) => o === option)}
+            onMouseEnter={() => setFocusedOptionIndex(index)}
+            focused={index === focusedOptionIndex}
+            aria-label={getOptionText(option)}
+          >
+            {renderOption ? renderOption(option) : getOptionText(option)}
+          </OptionItem>
+        ))}
+      </Options>
+    </Popup>
+  )
+}
+
+export default MultiSelect
+
+const SelectedValue = styled.div`
+  ${inputDefaultStyle(true, true, true)};
```

Comment:
I would like at some point to review our approach to inputs in general: text areas, input fields, select fields, etc. I feel that it could be done simpler.

Reply:
Agreed 💯